src/fmmax/vector.py

```diff
@@ -189,18 +189,26 @@ def _compute_tangent_field_no_batch(
     smoothness_loss_weight /= expansion.num_terms
 
     grid_shape: Tuple[int, int] = arr.shape[-2:]  # type: ignore[assignment]
-    arr = _filter_and_adjust_resolution(arr, expansion)
     grad = compute_gradient(arr, primitive_lattice_vectors)
+    grad = jnp.stack(
+        [
+            _filter_and_adjust_resolution(grad[..., 0], expansion),
+            _filter_and_adjust_resolution(grad[..., 1], expansion),
+        ],
+        axis=-1,
+    )
 
     # When the gradient is zero, we return a spatially invariant field, and provide a
-    # dummy gradient for the field calculation to avoid NaNs in gradient calculation.
+    # dummy gradient for the field calculation to avoid NaNs when backpropagating
+    # through the vector field calculation.
     gx_is_zero = jnp.all(
         jnp.isclose(grad[..., 0, jnp.newaxis], 0.0), axis=(-3, -2, -1), keepdims=True
     )
     gy_is_zero = jnp.all(
         jnp.isclose(grad[..., 1, jnp.newaxis], 0.0), axis=(-3, -2, -1), keepdims=True
     )
-    grad = jnp.where(gx_is_zero & gy_is_zero, jnp.ones_like(grad), grad)
+    dummy_grad = jnp.broadcast_to(jnp.asarray([1, 0], dtype=complex), grad.shape)
+    grad = jnp.where(gx_is_zero & gy_is_zero, dummy_grad, grad)
```

Comment:
The "double where" trick my be more canonical/robust here:

https://jax.readthedocs.io/en/latest/faq.html#gradients-contain-nan-where-using-where

Reply:
That's exactly what this is. :-)

We give a dummy gradient to the vector field calculation, so as to avoid it producing NaNs. This is the first `where`. Then, below we manually overwrite the result of that calculation with a safe value (with the second `where`).

Reply:
The issue previously was that grad was coming up as nonzero, and so the "safe" value was not used. This happened because the gradient was calculated for a "smoothed" field. Now we reverse the sequence --- first calculate the gradient for the raw permittivity, and then smooth this.

Reply:
Got it, thanks!